engine/src/board.rs

```diff
@@ -200,16 +213,39 @@ impl Board {
         };
     }
 
-    pub fn is_top_piece(&self, piece: &Piece, position: &Position) -> bool {
-        self.board.get(position).unwrap().last().unwrap() == piece
+    // Use this whenever you interact with untrusted input, e.g. user input, frontend input,
+    // history files, etc.
+    pub fn top_piece(&self, position: &Position) -> Result<Piece, GameError> {
+        if let Some(vec) = self.board.get(position) {
+            if let Some(piece) = vec.last() {
+                return Ok(*piece);
+            } else {
+                unreachable!("Found a key in board but no value");
+            }
+        } else {
+            return Err(GameError::NoPieceAtPosition {
+                position: position.to_string(),
+            });
+        }
+    }
+
+    // Only use this one when you can guarantee that the board state is correct
+    pub fn top_piece_trusted(&self, position: &Position) -> Piece {
+        *self
+            .board
+            .get(position)
+            .expect("The board should have had this position")
+            .last()
+            .expect("There should be a bug on this position")
     }
 
-    pub fn top_piece(&self, position: &Position) -> Piece {
-        *self.board.get(position).unwrap().last().unwrap()
+    pub fn is_top_piece(&self, piece: &Piece, position: &Position) -> Result<bool, GameError> {
+        let found = self.top_piece(position)?;
+        Ok(&found == piece)
     }
 
     pub fn top_bug(&self, position: &Position) -> Bug {
-        self.board.get(position).unwrap().last().unwrap().bug
+        self.top_piece_trusted(position).bug
```

Comment:
since this is a `pub` function, what's to stop a user of the library from calling this on an invalid position? shouldn't all `pub` functions be considered to have possible invalid input?